farming/logics/traits/farming.rs

```diff
@@ -64,13 +67,85 @@ pub trait Farming: Storage<Data> + Storage<ownable::Data> + FarmingGetters + Far
         self.data::<Data>().pool_info_length = pool_length
             .checked_add(1)
             .ok_or(FarmingError::AddOverflow2)?;
+
         self._emit_log_pool_addition_event(pool_length, alloc_point, lp_token, rewarder);
         Ok(())
     }
 
     #[ink(message)]
-    fn pending_arsw(&self, _pool_id: u32, _user: AccountId) -> Result<Balance, FarmingError> {
-        Ok(1_000_000_000_000_000_000)
+    #[modifiers(only_owner)]
+    fn set(
+        &mut self,
+        pool_id: u32,
+        alloc_point: u32,
+        rewarder: AccountId,
+        overwrite: bool,
+    ) -> Result<(), FarmingError> {
+        self._update_all_pools()?;
+        let pool_info = self
+            .get_pool_infos(pool_id)
+            .ok_or(FarmingError::PoolNotFound5)?;
+        self.data::<Data>().total_alloc_point = self
+            .data::<Data>()
+            .total_alloc_point
+            .checked_sub(pool_info.alloc_point)
+            .ok_or(FarmingError::SubUnderflow7)?
+            .checked_add(alloc_point)
+            .ok_or(FarmingError::AddOverflow9)?;
+
+        self.data::<Data>().pool_info.insert(
+            &pool_id,
+            &Pool {
+                alloc_point,
+                ..pool_info
+            },
+        );
+        let mut rewarder = rewarder;
+        if overwrite {
+            self.data::<Data>().rewarders.insert(&pool_id, &rewarder);
+            rewarder = self
+                .get_rewarder(pool_id)
+                .ok_or(FarmingError::RewarderNotFound)?;
+        }
```

Comment:
```suggestion
        if overwrite {
            self.data::<Data>().rewarders.insert(&pool_id, &rewarder);
        }
        let rewarder = self
            .get_rewarder(pool_id)
            .ok_or(FarmingError::RewarderNotFound)?;        
```
Seems correct.

Reply:
Solidity code is:
```
        if (overwrite) {
            rewarders[pid] = rewarder;
        }
        emit LogSetPool(
            pid,
            allocPoint,
            overwrite ? rewarder : rewarders[pid],
            overwrite
        );
```
Because of this line  `overwrite ? rewarder : rewarders[pid],` I am making it that way

Reply:
Yes, I think if overwrite is on, `self.data::<Data>().rewarders.insert(&pool_id, &rewarder)` will overwrite on chain data, thus 
```rust
        let rewarder = self
            .get_rewarder(pool_id)
            .ok_or(FarmingError::RewarderNotFound)?;  
```
is same as rewarder.
Otherwise, it will be just on-chain's rewarder. 

Thus, it corresponds to solidity's behavior.

When it comes to the original code, if overwrite is off, argument rewarder is set for event, which is not correct.